File: src/main/java/io/github/dqualizer/dqapi/controllers/rqa/RqaDefinitionController.java
package io.github.dqualizer.dqapi.controllers.rqa;

import io.github.dqualizer.dqapi.services.rqa.RqaDefinitionService;
import io.github.dqualizer.dqlang.types.rqa.definition.RuntimeQualityAnalysisDefinition;
import io.github.dqualizer.dqlang.types.rqa.definition.loadtest.LoadTestDefinition;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.net.URI;
import java.util.List;

@CrossOrigin(origins = "http://localhost:9090")
@RestController
@RequestMapping("/api/v2/rqa")
public class RqaDefinitionController {

    @Autowired
    private RqaDefinitionService service;

    @GetMapping
    public ResponseEntity<List<RuntimeQualityAnalysisDefinition>> getAllRqaDefinitions() {
        List<RuntimeQualityAnalysisDefinition> runtimeQualityAnalysisDefinitions = service.readAll();
        return new ResponseEntity<>(runtimeQualityAnalysisDefinitions, HttpStatus.OK);
    }

    @GetMapping("/{id}")
    public ResponseEntity<RuntimeQualityAnalysisDefinition> getRqaDefinitionById(@PathVariable String id){
        RuntimeQualityAnalysisDefinition rqaDefinition = service.readById(id);
        return new ResponseEntity<>(rqaDefinition, HttpStatus.FOUND);
    }

    @PostMapping
    public ResponseEntity<RuntimeQualityAnalysisDefinition> createRqaDefinition(@RequestBody RuntimeQualityAnalysisDefinition entity) {
        RuntimeQualityAnalysisDefinition runtimeQualityAnalysisDefinition = service.create(entity);
        URI location = ServletUriComponentsBuilder.fromCurrentRequest()
                .path("/{id}")
                .buildAndExpand(runtimeQualityAnalysisDefinition.getId())
                .toUri();

        return ResponseEntity.created(location).body(runtimeQualityAnalysisDefinition);
    }

    @PostMapping("/{rqaId}/loadtest")
    public ResponseEntity<RuntimeQualityAnalysisDefinition> createLoadtest(@PathVariable String rqaId, @RequestBody LoadTestDefinition entity) {

        RuntimeQualityAnalysisDefinition rqa = service.createLoadtest(rqaId, entity);

        URI location = ServletUriComponentsBuilder.fromCurrentRequest()
                .path(rqaId + "/loadtest/{id}")
                .buildAndExpand(rqa.getId())
                .toUri();

        return ResponseEntity.created(location).body(rqa);
    }
}

File: src/main/kotlin/io/github/dqualizer/dqapi/controllers/rqa/RqaDefinitionController.kt
package io.github.dqualizer.dqapi.controllers.rqa

import io.github.dqualizer.dqapi.models.loadtestDefinition.CreateLoadtestDefinitionDto
import io.github.dqualizer.dqapi.models.rqa.CreateRQADto
import io.github.dqualizer.dqapi.services.rqa.RqaDefinitionService
import io.github.dqualizer.dqlang.types.rqa.definition.RuntimeQualityAnalysisDefinition
import jakarta.servlet.http.HttpServletRequest
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.util.UriComponentsBuilder

@RestController
@RequestMapping("/api/v2/rqa")
class RqaDefinitionController(
  @Autowired val service: RqaDefinitionService,
  @Autowired val request: HttpServletRequest
) {
  @GetMapping
  fun readAll(): ResponseEntity<List<RuntimeQualityAnalysisDefinition>> {
    return ResponseEntity.ok(service.readAll())
  }

  @GetMapping("/{id}")
  fun readById(@PathVariable id: String): ResponseEntity<RuntimeQualityAnalysisDefinition> {
    return ResponseEntity.of(service.readById(id))
  }

  @PostMapping
  fun create(
    @RequestBody entity: CreateRQADto,
  ): ResponseEntity<RuntimeQualityAnalysisDefinition> {
    val runtimeQualityAnalysisDefinition = service.create(entity)

    val location = UriComponentsBuilder.fromUriString(request.requestURI)
      .path("/{id}")
      .buildAndExpand(runtimeQualityAnalysisDefinition.id)
      .toUri()

    return ResponseEntity.created(location).body(runtimeQualityAnalysisDefinition)
  }

  @PatchMapping("/{rqaId}/loadtest")
  fun updateLoadtest(
    @PathVariable rqaId: String,
    @RequestBody entity: CreateLoadtestDefinitionDto,
  ): ResponseEntity<RuntimeQualityAnalysisDefinition> {
    val loadTestDefinition = entity.build()
    val rqa = service.createLoadtest(rqaId, loadTestDefinition)

    val location = UriComponentsBuilder.fromUriString(request.requestURI)
      .path("/{id}")
      .buildAndExpand(rqa.id)
      .toUri()

    return ResponseEntity.created(location).body(rqa)
  }
}
